lib/withOnyx.tsx

```diff
@@ -0,0 +1,531 @@
+             */
+            checkEvictableKeys() {
+                // We will add this key to our list of recently accessed keys
+                // if the canEvict function returns true. This is necessary criteria
+                // we MUST use to specify if a key can be removed or not.
+                Object.keys(mapOnyxToState).forEach((propName) => {
+                    const mapping = mapOnyxToState[propName as keyof TOnyxProps] as WithOnyxMapping<TComponentProps, TOnyxProps>;
+
+                    if (mapping.canEvict === undefined) {
+                        return;
+                    }
+
+                    const canEvict = Str.result(mapping.canEvict as GenericFunction, this.props) as boolean;
+                    const key = Str.result(mapping.key as GenericFunction, this.props) as OnyxKey;
+
+                    if (!OnyxUtils.isSafeEvictionKey(key)) {
+                        throw new Error(`canEvict can't be used on key '${key}'. This key must explicitly be flagged as safe for removal by adding it to Onyx.init({safeEvictionKeys: []}).`);
+                    }
+
+                    if (canEvict) {
+                        OnyxUtils.removeFromEvictionBlockList(key, mapping.connectionID);
+                    } else {
+                        OnyxUtils.addToEvictionBlockList(key, mapping.connectionID);
+                    }
+                });
+            }
+
+            /**
+             * Takes a single mapping and binds the state of the component to the store
+             *
+             * @param mapping.key key to connect to. can be a string or a
+             * function that takes this.props as an argument and returns a string
+             * @param statePropertyName the name of the state property that Onyx will add the data to
+             * @param [mapping.initWithStoredValues] If set to false, then no data will be prefilled into the
+             *  component
+             * @param key to connect to Onyx with
+             */
+            connectMappingToOnyx(mapping: MapOnyxToState<TComponentProps, TOnyxProps>[keyof TOnyxProps], statePropertyName: keyof TOnyxProps, key: OnyxKey) {
+                const onyxMapping = mapOnyxToState[statePropertyName as keyof TOnyxProps] as WithOnyxMapping<TComponentProps, TOnyxProps>;
+
+                // Remember what the previous key was so that key changes can be detected when data is being loaded from Onyx. This will allow
+                // dependent keys to finish loading their data.
+                // eslint-disable-next-line no-param-reassign
+                onyxMapping.previousKey = key;
+
+                // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+                this.activeConnectionIDs[key] = Onyx.connect({
+                    ...mapping,
+                    key,
+                    statePropertyName: statePropertyName as string,
+                    withOnyxInstance: this as unknown as WithOnyxInstance,
```

Comment:
I think we should adjust `WithOnyxInstance` type. `withOnyx` === `WithOnyxInstance`:
```ts
class withOnyx extends React.Component<WithOnyxProps<TComponentProps, TOnyxProps>, WithOnyxState<TOnyxProps>> {

type WithOnyxInstance = Component<unknown, WithOnyxInstanceState<NullableKeyValueMapping>> & {
    setStateProxy: (cb: (state: Record<string, OnyxCollection<KeyValueMapping[OnyxKey]>>) => OnyxValue<OnyxKey>) => void;
    setWithOnyxState: (statePropertyName: OnyxKey, value: OnyxValue<OnyxKey>) => void;
};
```

It may not be a trivial thing to do, so if you think it's not worth the effort you can resolve the comment.

Reply:
Yeah I tried for a couple hours to get the typings working, but I had a lot of issues on `OnyxUtils`, not sure if it's worth it.. but I did some improvements